### PR TITLE
Refactor Cop#add_offense to use kwargs

### PR DESCRIPTION
**Pull request by akhramov**
At https://github.com/bbatsov/rubocop/pull/4717#issuecomment-328225732 @Drenmi suggested to turn `Cop::Cop#add_offense` arguments into keyword arguments, so we won't have to keep redundant `location` argument if we need to pass `message` argument.

This change mostly affects `InternalAffairs/OffenseLocationKeyword`, `InternalAffairs/RedundantLocationArgument`, `InternalAffairs/RedundantMessageArgument` cops, and also transforms all calls of `Cop::Cop#add_offense` with positional arguments to calls with
keyword arguments.

The subject of discussion, however, is whether we need to support the old method signature in order to ensure that we don't break any third-party cops.

Strategy might be as follows:
* Keep positional arguments in the method's signature
* Add a deprecation warning in case if someone still uses positional arguments. (any idea / guidance how to properly make a deprecation warning?)
* Create an internal affairs cop which enforces kwargs style.
* As far as we reach version X, remove the positional arguments version.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/internal_affairs/redundant_location_argument.rb [hunk 1]
```diff
@@ -10,36 +10,48 @@ module InternalAffairs
...
       class RedundantLocationArgument < Cop
         MSG = 'Redundant location argument to `#add_offense`.'.freeze
 
-        def_node_matcher :node_type_check, <<-PATTERN
-          (send nil :add_offense _ (sym :expression))
+        def_node_matcher :add_offense_kwargs, <<-PATTERN
+          (send nil :add_offense _ $hash)
+        PATTERN
+
+        def_node_matcher :location_expression_argument?, <<-PATTERN
```
Comment (Drenmi) [inline lib/rubocop/cop/internal_affairs/redundant_location_argument.rb]: Right now this is named after its implementation details. How about `#redundant_location_argument?`

File: lib/rubocop/cop/internal_affairs/redundant_location_argument.rb [hunk 2]
```diff
@@ -10,36 +10,48 @@ module InternalAffairs
...
+          (pair (sym :location) (sym :expression))
         PATTERN
 
         def on_send(node)
-          node_type_check(node) do
-            add_offense(node.last_argument)
-          end
+          offending_node = location_expression_argument(node)
```
Comment (Drenmi) [inline lib/rubocop/cop/internal_affairs/redundant_location_argument.rb]: I would make `#location_expression_argument` yield anything it finds, and `#add_offense` in a block.

File: lib/rubocop/cop/internal_affairs/redundant_message_argument.rb [hunk 3]
```diff
@@ -10,39 +10,61 @@ module InternalAffairs
...
       class RedundantMessageArgument < Cop
         MSG = 'Redundant message argument to `#add_offense`.'.freeze
 
         def_node_matcher :node_type_check, <<-PATTERN
-          (send nil :add_offense _offender _
-            {(const nil :MSG) (send nil :message) (send nil :message _offender)})
+          (send nil :add_offense $_node $hash)
         PATTERN
 
+        def_node_matcher :offending_message_argument, <<-PATTERN
```
Comment (Drenmi) [inline lib/rubocop/cop/internal_affairs/redundant_message_argument.rb]: Changing `offending` to `redundant` would provide some hint to *why* it is offending.

File: lib/rubocop/cop/internal_affairs/redundant_message_argument.rb [hunk 4]
```diff
@@ -10,39 +10,61 @@ module InternalAffairs
...
       #
       #   # good
-      #   add_offense(node, :expression)
-      #   add_offense(node, :expression, CUSTOM_MSG)
-      #   add_offense(node, :expression, message(other_node))
+      #   add_offense(node)
+      #   add_offense(node, message: CUSTOM_MSG)
+      #   add_offense(node, message: message(other_node))
       #
       class RedundantMessageArgument < Cop
```
Comment (Drenmi) [inline lib/rubocop/cop/internal_affairs/redundant_message_argument.rb]: Ugh. This implementation ended up with a lot of indirection, but I think our node pattern matchers currently can't match "any pair", so I might look into that. 🙂 For now, I think this implementation is okay.

File: lib/rubocop/cop/layout/indentation_width.rb [hunk 5]
```diff
@@ -237,10 +237,11 @@ def offense(body_node, indentation, style)
                  end
 
           indentation_name = style == 'normal' ? '' : "#{style} "
-          add_offense(node, offending_range(body_node, indentation),
-                      format("Use #{configured_indentation_width} (not %d) " \
-                             "spaces for #{indentation_name}indentation.",
-                             indentation))
+          msg = format("Use #{configured_indentation_width} (not %d) " \
```
Comment (Drenmi) [inline lib/rubocop/cop/layout/indentation_width.rb]: Feel free to extract this into a constant format string. 🙂

File: spec/rubocop/cop/internal_affairs/redundant_message_argument_spec.rb [hunk 6]
```diff
@@ -6,59 +6,72 @@
...
-                                     ^^^^^^^^^^^^^ Redundant message argument to `#add_offense`.
+      add_offense(node, message: message(node))
+                        ^^^^^^^^^^^^^^^^^^^^^^ Redundant message argument to `#add_offense`.
       RUBY
     end
 
     it 'auto-corrects' do
       new_source =
-        autocorrect_source('add_offense(node, :expression, message(node))')
+        autocorrect_source('add_offense(node, message: message(node))')
```
Comment (Drenmi) [inline spec/rubocop/cop/internal_affairs/redundant_message_argument_spec.rb]: We might want to test the auto-correct when there are additional arguments, to ensure that there are no syntax errors.

File: lib/rubocop/cop/internal_affairs/redundant_message_argument.rb (hunk 4 above)
Comment (akhramov) [reply lib/rubocop/cop/internal_affairs/redundant_message_argument.rb]: Can't agree more, we need "any pair" matchers desperately. 
If you're wondering, I have borrowed used pattern from `Rails/DelegateAllowBlank`:
https://github.com/bbatsov/rubocop/blob/7416bab77b9fd3c6f6476c71d5ecbd27910ec1a5/lib/rubocop/cop/rails/delegate_allow_blank.rb#L19-L25
https://github.com/bbatsov/rubocop/blob/7416bab77b9fd3c6f6476c71d5ecbd27910ec1a5/lib/rubocop/cop/rails/delegate_allow_blank.rb#L43-L47

I've looked inside `RuboCop::NodePattern` class, but didn't manage how to introduce the "any pair" matchers. Will invest some more time, though, will be happy if you or someone else will implement it earlier.

Comment (Drenmi) [reply lib/rubocop/cop/internal_affairs/redundant_message_argument.rb]: @akhramov: Yeah. The node pattern parser takes a while to wrap your head around. 🙂 The more people who are familiar with it, the better!

**Comment by yujinakayama** [comment]
> The subject of discussion, however, is whether we need to support the old method signature in order to ensure that we don't break any third-party cops.

I'm in favor of the support for backward compatibility.

**Comment by akhramov** [comment]
@yujinakayama / @Drenmi  I've added backwards compatibility support layer and a cop, which enforces kwargs style. Please see ca454b51289c25ab5f81f1e50445480508800967 for changes.

**Comment by bbatsov** [comment]
1. The build seems to be failing.
2. This needs to be rebased on top of `master`.

File: lib/rubocop/cop/cop.rb [hunk 7]
```diff
@@ -232,6 +222,43 @@ def custom_severity
...
+      def add_offense_deprecated(node, loc = :expression, message = nil,
+                                 severity = nil, &block)
+
+        message = <<-RUBY.strip_indent
+          Warning: The usage of positional location, message, and severity
+          parameters to #add_offense is deprecated.
+          Please use keyword arguments instead.
+
+          The positional arguments version of #add_offense will be removed in
+          Rubocop 0.52
```
Comment (bbatsov) [inline lib/rubocop/cop/cop.rb]: RuboCop ;-)

File: lib/rubocop/cop/cop.rb [hunk 8]
```diff
@@ -232,6 +222,43 @@ def custom_severity
           warn(Rainbow(message).red)
         end
       end
+
+      def add_offense_deprecated(node, loc = :expression, message = nil,
+                                 severity = nil, &block)
+
+        message = <<-RUBY.strip_indent
+          Warning: The usage of positional location, message, and severity
+          parameters to #add_offense is deprecated.
```
Comment (bbatsov) [inline lib/rubocop/cop/cop.rb]: Use "`Cop#add_offense`" here for clarity.

**Comment by akhramov** [comment]
@bbatsov, thanks, done :)

**Comment by bbatsov** [comment]
👍

**Pull request merged**